<a href="https://colab.research.google.com/github/endahwulandari2005-code/tugas-uts/blob/main/UTS_DATA_MINING.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##**1. Import Library**

In [18]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

tahap pertama yang dilakukan yaitu import library

import library digunakan dalam proses analisis data dan pembuatan model machine learning.

- `pandas` digunakan untuk membaca dan mengelola dataset.
- `numpy` digunakan untuk operasi numerik.
- `train_test_split` digunakan untuk membagi data menjadi data training dan validation.
- `RandomForestClassifier` digunakan untuk membangun model klasifikasi.
- `accuracy_score`, `confusion_matrix`, dan `classification_report` digunakan untuk mengevaluasi performa model.

##**2. Membaca Dataset**

In [19]:
train = pd.read_csv('data_training.csv')
test = pd.read_csv('data_testing.csv')

langkah selanjutnya yaitu membaca dataset

Dataset training dan testing dibaca menggunakan fungsi `read_csv()` dari library pandas.

- Dataset training digunakan untuk melatih model klasifikasi.
- Dataset testing digunakan untuk melakukan prediksi kualitas wine.

##**3. Melihat Data**

In [20]:
train.head()

,fixed acidity,volatile acidity,citric acid,residual sugar,chlorides,free sulfur dioxide,total sulfur dioxide,density,pH,sulphates,alcohol,quality,Id
0,7.3,0.740,0.08,1.7,0.094,10.0,45.0,0.99576,3.24,0.50,9.8,5,1366
1,8.1,0.575,0.22,2.1,0.077,12.0,65.0,0.99670,3.29,0.51,9.2,5,103
2,10.1,0.430,0.40,2.6,0.092,13.0,52.0,0.99834,3.22,0.64,10.0,7,942
3,12.9,0.500,0.55,2.8,0.072,7.0,24.0,1.00012,3.09,0.68,10.9,6,811
4,8.4,0.360,0.32,2.2,0.081,32.0,79.0,0.99640,3.30,0.72,11.0,6,918


Setelah membaca dataset, kita lihat datasetnya

Tahap ini dilakukan untuk melihat beberapa data teratas pada dataset training sehingga dapat diketahui struktur data dan variabel yang digunakan.

Dataset memiliki beberapa fitur kimiawi wine seperti:
- fixed acidity
- volatile acidity
- citric acid
- residual sugar
- chlorides
- alcohol
dan variabel target yaitu `quality`.

In [21]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 857 entries, 0 to 856
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         857 non-null    float64
 1   volatile acidity      857 non-null    float64
 2   citric acid           857 non-null    float64
 3   residual sugar        857 non-null    float64
 4   chlorides             857 non-null    float64
 5   free sulfur dioxide   857 non-null    float64
 6   total sulfur dioxide  857 non-null    float64
 7   density               857 non-null    float64
 8   pH                    857 non-null    float64
 9   sulphates             857 non-null    float64
 10  alcohol               857 non-null    float64
 11  quality               857 non-null    int64  
 12  Id                    857 non-null    int64  
dtypes: float64(11), int64(2)
memory usage: 87.2 KB




Berdasarkan hasil `train.info()`, diperoleh informasi bahwa:

- Dataset memiliki 857 baris data dan 13 kolom.
- Tidak terdapat missing value karena seluruh kolom memiliki jumlah non-null yang sama yaitu 857. Ini juga dikonfirmasi oleh hasil `train.isnull().sum()` yang menunjukkan nol missing value untuk setiap kolom.
- Sebagian besar variabel bertipe data `float64` karena berisi nilai numerik kontinu.
- Variabel `quality` dan `Id` bertipe `int64`.

Karena tidak terdapat missing value, maka data dapat langsung digunakan pada tahap pemodelan.

##**4. Mengecek Missing Value**

In [22]:
train.isnull().sum()

,0
fixed acidity,0
volatile acidity,0
citric acid,0
residual sugar,0
chlorides,0
free sulfur dioxide,0
total sulfur dioxide,0
density,0
pH,0
sulphates,0


Meskipun dalam hasil train.info() menunjukkan tidak ada missing value, tapi kita juga butuh konfirmasi oleh hasil `train.isnull().sum()` yang menunjukkan nol missing value untuk setiap kolom.


- `train.isnull()` akan mengembalikan DataFrame boolean dengan `True` di setiap lokasi di mana ada nilai null dan `False` di tempat lain.
- `.sum()` kemudian akan menghitung jumlah `True` untuk setiap kolom, yang secara efektif memberikan jumlah total nilai null per kolom.

Output dari `train.isnull().sum()` menunjukkan bahwa tidak ada kolom yang memiliki nilai null, karena semua hitungan adalah 0. Ini mengkonfirmasi bahwa dataset `train` bersih dari missing values dan siap untuk digunakan dalam pemodelan.

##**5. Menentukan Fitur dan Target**

misalkan targetnya adalah quality

In [23]:
X = train.drop('quality', axis=1)
y = train['quality']

Selanjutnya kita menentukan fitur dan target

Pada tahap ini, kita akan memisahkan dataset menjadi dua bagian utama: fitur (features) dan variabel target (target variable).

- **Fitur (X)**: Ini adalah variabel-variabel independen yang akan digunakan untuk membuat prediksi. Dalam kasus ini, kita mengambil semua kolom dari DataFrame `train` kecuali kolom `'quality'`. Metode `drop('quality', axis=1)` digunakan untuk menghapus kolom 'quality' dari `train`, dengan `axis=1` menandakan bahwa kita menghapus kolom, bukan baris.

- **Variabel Target (y)**: Ini adalah variabel dependen yang ingin kita prediksi. Dalam kasus ini, variabel target adalah `'quality'`, yang kita ambil dari kolom 'quality' dari DataFrame `train`.

##**6. Membagi Data Training dan Validation**

In [24]:
X_train, X_valid, y_train, y_valid = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42
)

Setelah memisahkan fitur dan target, langkah selanjutnya adalah membagi dataset ini menjadi data training (pelatihan) dan data validation (validasi). Tujuan dari pembagian ini adalah untuk melatih model pada sebagian data dan kemudian menguji kinerjanya pada data yang belum pernah dilihat oleh model sebelumnya, untuk memastikan model dapat digeneralisasi dengan baik dan tidak *overfitting*.

Fungsi `train_test_split` dari `sklearn.model_selection` digunakan untuk melakukan pembagian ini dengan parameter:

- `X`: Fitur-fitur dari dataset.
- `y`: Variabel target dari dataset.
- `test_size=0.2`: Menentukan proporsi data yang akan digunakan sebagai data validasi. Dalam kasus ini, 20% dari total data akan menjadi data validasi, dan sisanya (80%) akan menjadi data training.
- `random_state=42`: Parameter ini digunakan untuk mengontrol pengacakan data. Dengan menetapkan nilai integer (misalnya 42), kita memastikan bahwa setiap kali kode dijalankan, pembagian data akan selalu sama, sehingga hasil yang diperoleh bisa direproduksi (konsisten).

Hasil dari pembagian ini adalah empat subset data:

- `X_train`: Fitur-fitur untuk pelatihan model.
- `X_valid`: Fitur-fitur untuk validasi model.
- `y_train`: Variabel target untuk pelatihan model.
- `y_valid`: Variabel target untuk validasi model.

##**7. Membuat Model (Random Forest)**

In [25]:
model = RandomForestClassifier(
    n_estimators=100,
    random_state=42
)

model.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

Pada tahap ini, kita akan membangun model klasifikasi menggunakan algoritma Random Forest.

**Random Forest** adalah algoritma pembelajaran mesin yang kuat dan serbaguna yang bekerja dengan membangun banyak pohon keputusan selama pelatihan dan mengeluarkan kelas yang merupakan modus (mode) dari kelas-kelas keluaran oleh masing-masing pohon.

Berikut adalah penjelasan parameter yang digunakan:

- `RandomForestClassifier()`: Ini adalah kelas untuk membangun model Random Forest untuk tugas klasifikasi.
- `n_estimators=100`: Parameter ini menentukan jumlah pohon keputusan yang akan dibangun dalam *forest*. Semakin banyak pohon, semakin kuat model, tetapi juga semakin lama waktu komputasi. Nilai 100 adalah titik awal yang umum.
- `random_state=42`: Parameter ini digunakan untuk mengontrol pengacakan dalam proses pembuatan pohon. Dengan menetapkan nilai integer (misalnya 42), kita memastikan bahwa hasil pembangunan model akan selalu sama setiap kali kode dijalankan, sehingga hasilnya dapat direproduksi.

Setelah model diinisialisasi, metode `model.fit(X_train, y_train)` digunakan untuk melatih model. Pada langkah ini, model mempelajari pola dari data pelatihan (`X_train` sebagai fitur dan `y_train` sebagai target) untuk membuat keputusan klasifikasi di masa mendatang.

##**8. Prediksi Data**

In [26]:
y_pred = model.predict(X_valid)

Setelah model dilatih menggunakan data training, langkah selanjutnya adalah menggunakan model tersebut untuk membuat prediksi. Pada tahap ini, kita akan melakukan prediksi pada data validasi (`X_valid`) yang belum pernah dilihat oleh model selama proses pelatihan.

- `model.predict(X_valid)`: Metode `predict()` dari objek model Random Forest (`model`) digunakan untuk menghasilkan label kelas prediksi (`y_pred`) berdasarkan fitur-fitur dari data validasi (`X_valid`). Hasil dari prediksi ini adalah serangkaian nilai yang menunjukkan kualitas wine yang diprediksi oleh model untuk setiap sampel di set validasi.

##**9. Evaluasi Model**

Setelah model dilatih dan digunakan untuk membuat prediksi, langkah krusial berikutnya adalah mengevaluasi seberapa baik kinerja model tersebut. Evaluasi model membantu kita memahami kekuatan dan kelemahan model dalam memprediksi kelas target. Ada beberapa metrik umum yang digunakan:

In [27]:
#Accuracy
accuracy = accuracy_score(y_valid, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.5930232558139535


Accuracy digunakan untuk mengukur tingkat ketepatan prediksi model

Output `Accuracy: 0.5930232558139535` menunjukkan bahwa model kita memiliki tingkat akurasi sekitar **59.3%** pada data validasi. Ini berarti sekitar 59.3% dari semua prediksi yang dibuat oleh model untuk kualitas wine pada set validasi adalah benar.

In [28]:
#Confusion Matrix
print(confusion_matrix(y_valid, y_pred))

[[ 0  1  2  0  0]
 [ 0 47 20  0  0]
 [ 0 23 46  9  0]
 [ 0  1 11  9  0]
 [ 0  0  0  3  0]]


Confusion matrix digunakan untuk melihat jumlah prediksi benar dan salah pada masing-masing kelas.

Output matriks konfusi:
```
[[ 0  1  2  0  0]
 [ 0 47 20  0  0]
 [ 0 23 46  9  0]
 [ 0  1 11  9  0]
 [ 0  0  0  3  0]]
```
Matriks ini menunjukkan perbandingan antara nilai sebenarnya (`y_valid`, baris) dengan nilai prediksi (`y_pred`, kolom). Angka-angka dalam laporan klasifikasi (di bawah) mengindikasikan bahwa kelas kualitas yang ada kemungkinan adalah 4, 5, 6, 7, dan 8.

Mari kita coba interpretasikan (asumsi urutan kelas adalah 4, 5, 6, 7, 8):

- **Baris 1 (Kualitas Sebenarnya 4):** `[0 1 2 0 0]`
  - Tidak ada kualitas 4 yang diprediksi benar (0 TP).
  - 1 kualitas 4 diprediksi sebagai 5.
  - 2 kualitas 4 diprediksi sebagai 6.
  - Total ada 3 sampel kualitas 4 (lihat `support` di laporan klasifikasi).

- **Baris 2 (Kualitas Sebenarnya 5):** `[0 47 20 0 0]`
  - 47 kualitas 5 diprediksi benar (47 TP).
  - 20 kualitas 5 diprediksi sebagai 6.
  - Total ada 67 sampel kualitas 5.

- **Baris 3 (Kualitas Sebenarnya 6):** `[0 23 46 9 0]`
  - 46 kualitas 6 diprediksi benar (46 TP).
  - 23 kualitas 6 diprediksi sebagai 5.
  - 9 kualitas 6 diprediksi sebagai 7.
  - Total ada 78 sampel kualitas 6.

- **Baris 4 (Kualitas Sebenarnya 7):** `[0 1 11 9 0]`
  - 9 kualitas 7 diprediksi benar (9 TP).
  - 1 kualitas 7 diprediksi sebagai 5.
  - 11 kualitas 7 diprediksi sebagai 6.
  - Total ada 21 sampel kualitas 7.

- **Baris 5 (Kualitas Sebenarnya 8):** `[0 0 0 3 0]`
  - Tidak ada kualitas 8 yang diprediksi benar (0 TP).
  - 3 kualitas 8 diprediksi sebagai 7.
  - Total ada 3 sampel kualitas 8.

Matriks ini menunjukkan bahwa model cenderung salah memprediksi kualitas 4 dan 8, serta sering bingung antara kualitas 5, 6, dan 7.

In [29]:
#Classification Report
print(classification_report(y_valid, y_pred))

              precision    recall  f1-score   support

           4       0.00      0.00      0.00         3
           5       0.65      0.70      0.68        67
           6       0.58      0.59      0.59        78
           7       0.43      0.43      0.43        21
           8       0.00      0.00      0.00         3

    accuracy                           0.59       172
   macro avg       0.33      0.34      0.34       172
weighted avg       0.57      0.59      0.58       172



/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


Classification report digunakan untuk melihat nilai precision, recall, dan f1-score.

Output laporan klasifikasi memberikan metrik `precision`, `recall`, dan `f1-score` untuk setiap kelas kualitas (4, 5, 6, 7, 8), serta `accuracy`, `macro avg`, dan `weighted avg` secara keseluruhan.

```
              precision    recall  f1-score   support

           4       0.00      0.00      0.00         3
           5       0.65      0.70      0.68        67
           6       0.58      0.59      0.59        78
           7       0.43      0.43      0.43        21
           8       0.00      0.00      0.00         3

    accuracy                           0.59       172
   macro avg       0.33      0.34      0.34       172
weighted avg       0.57      0.59      0.58       172
```

**Interpretasi per Kelas:**

- **Kelas 4:**
  - `precision`, `recall`, `f1-score`: 0.00. Ini sangat rendah karena model tidak mampu memprediksi satu pun sampel kualitas 4 dengan benar (berdasarkan matriks konfusi, semua kualitas 4 diprediksi sebagai 5 atau 6). `support` menunjukkan ada 3 sampel kualitas 4 yang sebenarnya.
- **Kelas 5:**
  - `precision`: 0.65 (65% dari prediksi 'kualitas 5' adalah benar-benar kualitas 5).
  - `recall`: 0.70 (model berhasil mengidentifikasi 70% dari semua kualitas 5 yang sebenarnya).
  - `f1-score`: 0.68. `support`: 67 sampel.
- **Kelas 6:**
  - `precision`: 0.58.
  - `recall`: 0.59.
  - `f1-score`: 0.59. `support`: 78 sampel (kelas mayoritas).
- **Kelas 7:**
  - `precision`: 0.43.
  - `recall`: 0.43.
  - `f1-score`: 0.43. `support`: 21 sampel.
- **Kelas 8:**
  - `precision`, `recall`, `f1-score`: 0.00. Sama seperti kelas 4, model tidak dapat memprediksi kelas ini dengan baik. `support`: 3 sampel.

**Metrik Keseluruhan:**

- `accuracy`: 0.59. Ini adalah akurasi total model yang sudah kita lihat sebelumnya.
- `macro avg`: Rata-rata sederhana (unweighted) dari presisi, recall, dan f1-score di semua kelas. Nilainya rendah (0.33-0.34) karena kinerja buruk pada kelas 4 dan 8 sangat menurunkannya.
- `weighted avg`: Rata-rata dari presisi, recall, dan f1-score, tetapi dibobot berdasarkan jumlah sampel (`support`) di setiap kelas. Nilainya (0.57-0.59) lebih dekat ke akurasi karena kelas mayoritas (5 dan 6) memiliki bobot lebih besar dan kinerjanya lebih baik. `Weighted avg` lebih relevan ketika kelas tidak seimbang.

Secara keseluruhan, laporan ini menunjukkan bahwa model kesulitan membedakan antara kualitas wine, terutama pada kelas minoritas (4 dan 8), yang mendapatkan skor sangat rendah. Model paling baik dalam memprediksi kualitas 5 dan 6, tetapi masih ada banyak ruang untuk perbaikan.

##**10. Prediksi Data Testing**

In [30]:
test_pred = model.predict(test)

Setelah model dilatih dan dievaluasi pada data validasi, langkah selanjutnya adalah menggunakan model yang telah dibangun untuk membuat prediksi pada dataset `test` yang sebenarnya. Dataset `test` ini mewakili data baru yang belum pernah dilihat model sama sekali.

- `test_pred = model.predict(test)`: Baris kode ini menggunakan metode `predict()` dari objek model Random Forest (`model`) untuk menghasilkan prediksi kelas kualitas wine untuk setiap baris data di dataset `test`. Hasilnya disimpan dalam variabel `test_pred`, yang merupakan array NumPy berisi nilai kualitas wine yang diprediksi.

##**11. Membuat file CSV hasil prediksi**

In [31]:
submission = pd.DataFrame({
    'Id': test['Id'],
    'quality': test_pred
})

In [35]:
submission = pd.DataFrame({
    'Id': test['Id'],
    'quality': test_pred
})

Sesuai perintah kita buat file berbentuk CSV hasil prediksi

- `submission = pd.DataFrame({'Id': range(len(test_pred)), 'quality': test_pred})`: Kode ini membuat DataFrame `submission` dengan kolom 'Id' (ID berurutan) dan 'quality' (hasil prediksi dari `test_pred`). Ini adalah format standar untuk menyimpan hasil prediksi.

##**12. Simpan CSV**

In [33]:
submission.to_csv('hasilprediksi.csv', index=False)

In [34]:
from google.colab import files
files.download('hasilprediksi.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Dan yang terakhir yaitu Menyimpan dan Mengunduh File CSV

- `submission.to_csv('hasilprediksi.csv', index=False)`: Kode ini menyimpan DataFrame `submission` ke file `hasilprediksi.csv`. `index=False` mencegah penulisan indeks DataFrame ke dalam file CSV.
- `files.download('hasilprediksi.csv')`: Kode ini mengunduh file CSV yang telah dibuat langsung ke perangkat Anda.